File: language-support/js/daml-ledger/src/mapping/completion_stream_response.ts
// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

import * as mapping from '.';
import * as ledger from '..';
import * as grpc from 'daml-grpc';

export const CompletionStreamResponse: mapping.Mapping<grpc.CompletionStreamResponse, ledger.CompletionStreamResponse> = {
    toObject(message: grpc.CompletionStreamResponse): ledger.CompletionStreamResponse {
        const response: ledger.CompletionStreamResponse = {};
        const completions = message.getCompletionsList();
        if (completions) {
            response.completions = completions.map(c => mapping.Completion.toObject(c));
        }
        if (message.hasCheckpoint()) {
            response.checkpoint = mapping.Checkpoint.toObject(message.getCheckpoint()!)
        }
        return response;
    },
    toMessage(object: ledger.CompletionStreamResponse): grpc.CompletionStreamResponse {
        const result = new grpc.CompletionStreamResponse();
        if (object.checkpoint) {
            result.setCheckpoint(mapping.Checkpoint.toMessage(object.checkpoint));
        }
        if (object.completions) {
            result.setCompletionsList(object.completions.map((c: ledger.Completion) => mapping.Completion.toMessage(c)));
        }
        return result;
    }
}

File: daml/language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/CompletionStreamResponse.java
// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.daml.ledger.javaapi.data;

import com.daml.ledger.api.v1.CommandCompletionServiceOuterClass;
import com.daml.ledger.api.v1.CompletionOuterClass;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import org.checkerframework.checker.nullness.qual.NonNull;

public class CompletionStreamResponse {

  private final Optional<Checkpoint> checkpoint;

  private final List<CompletionOuterClass.Completion> completions;

  public CompletionStreamResponse(
      @NonNull Optional<Checkpoint> checkpoint,
      @NonNull List<CompletionOuterClass.@NonNull Completion> completions) {
    this.checkpoint = checkpoint;
    this.completions = completions;
  }

  public static CompletionStreamResponse fromProto(
      CommandCompletionServiceOuterClass.CompletionStreamResponse response) {
    if (response.hasCheckpoint()) {
      Checkpoint checkpoint = Checkpoint.fromProto(response.getCheckpoint());
      return new CompletionStreamResponse(Optional.of(checkpoint), response.getCompletionsList());
    } else {
      return new CompletionStreamResponse(Optional.empty(), response.getCompletionsList());
    }
  }

  public CommandCompletionServiceOuterClass.CompletionStreamResponse toProto() {
    CommandCompletionServiceOuterClass.CompletionStreamResponse.Builder builder =
        CommandCompletionServiceOuterClass.CompletionStreamResponse.newBuilder();
    this.checkpoint.ifPresent(c -> builder.setCheckpoint(c.toProto()));
    builder.addAllCompletions(this.completions);
    return builder.build();
  }

  @NonNull
  public Optional<Checkpoint> getCheckpoint() {
    return checkpoint;
  }

  @NonNull
  public List<CompletionOuterClass.@NonNull Completion> getCompletions() {
    return completions;
  }

  @Override
  public String toString() {
    return "CompletionStreamResponse{"
        + "checkpoint="
        + checkpoint
        + ", completions="
        + completions
        + '}';
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    CompletionStreamResponse that = (CompletionStreamResponse) o;
    return Objects.equals(checkpoint, that.checkpoint)
        && Objects.equals(completions, that.completions);
  }

  @Override
  public int hashCode() {

    return Objects.hash(checkpoint, completions);
  }
}
